File: 015_Binary_Search_Advanced_Problems/Aggressive_Cows.swift
import Foundation

class Solution {
    
    func isPossible(stalls: [Int], minDist: Int, k: Int) -> Bool {
        var cows = 1
        var lastCowPosition = stalls[0]
        
        for i in 1..<stalls.count {
            if stalls[i] - lastCowPosition >= minDist {
                cows += 1
                lastCowPosition = stalls[i]
                if cows >= k {
                    return true
                }
            }
        }
        return false
    }

    func solve(n: Int, k: Int, stalls: [Int]) -> Int {
        let sortedStalls = stalls.sorted()
        var low = 1
        var high = sortedStalls.last! - sortedStalls.first!
        var res = 0
        
        while low <= high {
            let mid = (low + high) / 2
            if isPossible(stalls: sortedStalls, minDist: mid, k: k) {
                res = mid
                low = mid + 1
            } else {
                high = mid - 1
            }
        }
        return res
    }
}

// Input
let t = 1
let n = 5
let k = 3
let stalls = [1, 2, 8, 4, 9]

let solution = Solution()
let result = solution.solve(n: n, k: k, stalls: stalls)
print(result)

File: 015_Binary_Search_Advanced_Problems/Aggressive_Cows.py
class Solution:
    def is_possible(self, stalls, min_dist, k):
        cows = 1
        last_cow_position = stalls[0]

        for i in range(1, len(stalls)):
            if stalls[i] - last_cow_position >= min_dist:
                cows += 1
                last_cow_position = stalls[i]
                if cows >= k:
                    return True
        return False

    def solve(self, n, k, stalls):
        stalls.sort()
        low, high = 1, stalls[-1] - stalls[0]
        res = 0

        while low <= high:
            mid = (low + high) // 2
            if self.is_possible(stalls, mid, k):
                res = mid
                low = mid + 1
            else:
                high = mid - 1
        return res


# Input handling
t = int(input().strip())

for _ in range(t):
    n, k = map(int, input().strip().split())
    stalls = list(map(int, input().strip().split()))

    obj = Solution()
    result = obj.solve(n, k, stalls)
    print(result)


'''
example input
    2         : t
    5 3       : n ,k
    1 2 8 4 9 : stalls

example output
    3

'''
